### PR TITLE
Add pre/postEventFunction to addDebugHook

### PR DESCRIPTION
**Pull request by Necktrox**
This PR adds, see title, two new debug hook types called **preEventFunction** and **postEventFunction**. You can now have greater control over event handling throughout your entire server if you want to debug your own implemented functionality in this region.

Wiki: [addDebugHook](https://wiki.multitheftauto.com/wiki/AddDebugHook)
```lua
string preEventFunction( resource eventResource, string eventName, element eventSource, element eventClient, string eventFilename, int eventLineNumber, resource functionResource, string functionFilename, int functionLineNumber, ...eventArgs )
       postEventFunction( resource eventResource, string eventName, element eventSource, element eventClient, string eventFilename, int eventLineNumber, resource functionResource, string functionFilename, int functionLineNumber, ...eventArgs )
```

### CODE REVIEW

File: Shared/mods/deathmatch/logic/CDebugHookManager.cpp [hunk 1]
```diff
@@ -132,7 +136,7 @@ bool CDebugHookManager::RemoveDebugHook( EDebugHookType hookType, const CLuaFunc
 ///////////////////////////////////////////////////////////////
 void CDebugHookManager::OnLuaMainDestroy( CLuaMain* pLuaMain )
 {
-    for( uint hookType = EDebugHook::PRE_EVENT ; hookType <= EDebugHook::POST_FUNCTION ; hookType++ )
+    for( uint hookType = EDebugHook::PRE_EVENT ; hookType <= EDebugHook::POST_EVENT_FUNCTION ; hookType++ )
```
Comment (qaisjp) [inline Shared/mods/deathmatch/logic/CDebugHookManager.cpp]: Could `DUMMY_LAST_DEBUG_HOOK` be used here, and defined where the enums are defined?

File: Client/mods/deathmatch/logic/CMapEventManager.cpp [hunk 2]
```diff
@@ -193,6 +193,9 @@ bool CMapEventManager::Call ( const char* szName, const CLuaArguments& Arguments
                     if ( bEnabled )
                         g_pCore->LogEvent ( 0, "Lua Event", pMapEvent->GetVM ()->GetScriptName (), szName );
 
+                    if ( !g_pClientGame->GetDebugHookManager()->OnPreEventFunction( szName, Arguments, pSource, NULL, pMapEvent ) )
```
Comment (qaisjp) [inline Client/mods/deathmatch/logic/CMapEventManager.cpp]: is `nullptr` possible?

**Review by qaisjp — COMMENTED** [review]
Nice stuff. I've suggested a thing or two, but if you've given this thorough testing, feel free to adjust the labels and give this a merge 🚢 🇮🇹 (:shipit:)

**Comment by Necktrox** [comment]
I don't have any usecase for this anymore personally. Does anybody need this? If not, then there is no reason to merge this.

**Comment by qaisjp** [comment]
What was your use-case?

**Comment by Necktrox** [comment]
To measure CPU usage for onClientRender per resource/eventhandler, which you can't do with onPreEvent onPostEvent.

**Comment by AlexTMjugador** [comment]
I feel that giving scripters and server owners profiling tools like this can't hurt, neither be a bloat, but indeed it can be something useful to gather information about how are concrete parts of a resource behaving.

Talking about the particular use-case that @Necktrox mentioned, I think that onClientRender can be a major performance bottleneck for the server when not so well designed algorithms are in place or just when a need to do extremely frequently operations arises (depending on the gamemode, it may be necessary to tie some logic at the execution speed of the game. And it could happen that one big resource does those operations). One could argue that in well designed scripts that shouldn't be a performance hog, and it would be right. But at least this PR would provide people who design that performance hogs tools to prove that they are doing something wrong, right?

Besides, CPU usage can be a [concern for server owners](https://stackoverflow.com/a/28309057) (albeit unlikely sor a MTA server alone), so anything that could give them more factual data about how are the CPU cycles being used seems welcome.

**Comment by Necktrox** [comment]
What's possible:
[pre/postEventFunction](https://hastebin.com/venoxekasa.swift)
[pre/postEvent](https://hastebin.com/nonogonewe.swift)
[pre/postFunction](https://hastebin.com/olivipiwip.swift)

Same examples, but hosted on GitHub Gist:
[pre/postEventFunction](https://gist.github.com/Necktrox/72cf3b0cd17bccf670fb1e2cf930bd92)
[pre/postEvent](https://gist.github.com/Necktrox/ab23c4bec5f08a454ea6ba4ae44f88da)
[pre/postFunction](https://gist.github.com/Necktrox/fae1cc41475c3ab571d5dc7743a309d7)

**Pull request merged**